rfcs/QueryLevelNullability.md

```diff
@@ -0,0 +1,269 @@
+        /// Since the `Business.name` schema field is nullable, the corresponding codegen Swift property is `Optional`
+        let name: String?
+      }
+    }
+  }
+  ```
+  The query can then be fetched, and the resulting data handled, as follows:
+  ```swift
+  GraphQL.fetch(query: GetBusinessNameQuery(id: "foo"), completion: { result in
+    guard case let .success(gqlResult) = result, let business = gqlResult.data?.business else { return }
+
+    // Often, the client needs to provide a default value in case `name` is `null`.
+    print(business?.name ?? "null")
+  }
+  ```
+
+## 📜 Problem Statement
+
+The two modern languages used on mobile clients, Swift and Kotlin, are both non-null by default. By contrast, in a
+GraphQL type system, every field is nullable by default. From the [official GraphQL best practice](https://graphql.org/learn/best-practices/#nullability):
+
+> This is because there are many things that can go awry in a networked service backed by databases and other
+> services. A database could go down, an asynchronous action could fail, an exception could be thrown. Beyond
+> simply system failures, authorization can often be granular, where individual fields within a request can
+> have different authorization rules.
+
+This discrepancy means that developers using strongly-typed languages must perform `null` checks anywhere that GraphQL
+codegen types are used, significantly decreasing the benefits of codegen. In some cases, the codegen types may be so
+complex that a new model type which encapsulates the `null` handling is written manually.
+
+While the schema can have nullable fields for valid reasons (such as federation), in some cases the client wants to decide
+if it accepts a `null` value for the result to simplify the client-side logic. In addition, syntax for this concept
+would allow codegen tooling to generate model types that are more ergonomic to work with, since the since the model
+type's properties would have the desired nullability.
+
+## 🧑‍💻 Proposed syntax
+
+The client can express that a schema field is required by using the `!` syntax in the query definition:
+```graphql
+query GetBusinessName($id: String!) {
+  business(id: $id) {
+    name!
+  }
+}
+```
+Semantically the GraphQL `!` operator is nearly identical to its counterpart in Swift (also represented by `!`) which is
+referred to as the "force unwrap operator".
+
+In Swift, for example, you can cast a string to an integer with `Int("5")` but the string being cast may not be a valid
+number, so that statement will evaluate to `null` rather than an integer if the string cannot be cast to an integer. If
+you want to ensure that the statement does not return `null` you can instead write `Int("5")!`. If you do that, an
+exception will be thrown if the statement would evaluate to `null`.
+
+In GraphQL, the `!` operator will act similarly. In the case that `name` does not exist, the query will return an error
+to the client rather than any data.
```

Comment:
Can you expand here? My reading of this is that if the value of a field with `!` is null at runtime, then the GraphQL executor must treat that identically to the case that the field resolver raised an exception, including returning an entry in the top-level `errors` list and nulling out the parent field. 

This would mean that in a query such as the following, `other_media_field` being null at runtime would cause the entire query to return effectively no data (just `{data: {media: null}, errors: [...]}`). This would prevent any part of the UI from rendering because a single field was missing for one part:

```
query MediaQuery {
  media(id: "...") {
     name
     thumbnail
     ...MediaFragment
  }
}
fragment MediaFragment on Media {
  other_media_field @requried
}
```

How are you thinking about handling cases such as this?

Reply:
I read this as any data that is `null` and has a `!` on the field in the query would cause the whole query to return an error and no data. Even if that field is in some named fragment.

I think this would limit the usefulness extremely, basically to very simple queries. To take Yelp as an example, I don't think it'd be a good idea to error a search query if a single search result item failed to generate a star rating (assuming that's required to render a search result item).

It would be good to clarify what the behavior is if the `!` is in a named fragment.

Reply:
At Netflix, we have implemented this on the server (using a directive syntax) and defined it to have identical semantics as the existing SDL [Non-Null behavior](https://spec.graphql.org/June2018/#sec-Errors-and-Non-Nullabilit). This has worked well.

I've proposed a [clarification](https://github.com/twof/graphql-spec/pull/5) to specify this.

Reply:
@fotoetienne Thanks for adding the clarification, that confirms my interpretation and brings us back to my original question:

How are you thinking about the fact that a single fragment with a `!` can cause data for an entire request to be nulled out? As @captbaritone mentioned above, one of our goals / requirements when designing `@required` was that it is absolutely critical that a "missing" required value for one component of a UI cannot cause data for other parts of the UI to go missing. The design here violates that constraint.

Reply:
@josephsavona That's a great question. Our current server logic will make a field act as Non-Null is there _any_ Non-Null designation in the query. This query example:

```graphql
fragment Preview on Media {
  thumbnail(size: 32)

  ... on Photo {
    thumbnail(size: 32) @nonNull
  }
}
```

would essentially be interpreted by the server as:

```graphql
fragment Preview on Media {
  thumbnail(size: 32) @nonNull

  ... on Photo {
    thumbnail(size: 32) @nonNull
  }
}
```

It is at least non-ambiguous, but I can see how it is a challenge for fragment modularity.

Reply:
@fotoetienne Thanks for clarifying. That seems like a severe limitation that would make applications significantly less robust than they can and arguably should be. How do you manage this in practice? Do you encourage fields with `!` to be given highly unique aliases so that you prevent removing data for other parts of a UI?

Reply:
Short answer - it hasn't come up _for us_ yet, but it's an important edge-case to consider.

Long answer - Fragment modularity is particularly crucial clients that rely heavily on components tied to individually defined Fragments that are compiled into large queries. We don't currently have many teams using that approach.

I agree that this is an important edge-case that we should think through some more. We want a design that will lead UI developers to fall into "the pit of success" rather than allowing surprising behavior like you've described.

Reply:
An initial thought, probably flawed: reversing the server-defined logic might help. Only honor nonNull if _all_ references to that field specify it. This would, however, require client codegen to be aware of an entire query and only make types nonNull if all fragments in the query specify nonNull.

Reply:
Making conflicting definitions between fragments be a validation error (like you did with @requires) is the safest approach. I like that. It's probably an acceptable limitation that you just can't have differing nullability for the same field.

Reply:
@joesavona even unique aliases would still null out the object by bubbling up to the parent. It's really the same as any specific field being non-nullable in the server schema and ending up null which bubbles up.